sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -623,8 +624,14 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val eval = child.genCode(ctx)
     val nullSafeCast = nullSafeCastFunction(child.dataType, dataType, ctx)
+
+    // Below the code comment including `eval.value` and `eval.isNull` is a trick. It makes the two
+    // expr values are referred by this code block.
     ev.copy(code = eval.code +
-      castCode(ctx, eval.value, eval.isNull, ev.value, ev.isNull, dataType, nullSafeCast))
+      code"""
+        // Cast from ${eval.value}, ${eval.isNull}
+        ${castCode(ctx, eval.value, eval.isNull, ev.value, ev.isNull, dataType, nullSafeCast)}
```

Comment:
I might miss something, but don't we need to pass `ExprValue` or `Block` and return `Block` from `castCode()` to keep track of code snippets and inputs?
We can say the same thing anywhere we create a smaller code snippet using `s` interpolation.

Reply:
We can use `ExprValue` and `Block` around everywhere in `Cast` here (I did in a local branch), just it increases some code diff, so I'm not sure if we want it be here or as a follow-up if it's easier for review.

Inputs to `Cast` should be only `eval.value` and `eval.isNull` as it's only child expression to `Cast`. We add generated codes here but they are not actually input expressions to `Cast`. So for now I manually let `eval.value` and `eval.isNull` be tracked.